src/Components/Server/src/Circuits/CircuitHost.cs

```diff
@@ -133,7 +134,71 @@ internal void SendPendingBatches()
             // Dispatch any buffered renders we accumulated during a disconnect.
             // Note that while the rendering is async, we cannot await it here. The Task returned by ProcessBufferedRenderBatches relies on
             // OnRenderCompleted to be invoked to complete, and SignalR does not allow concurrent hub method invocations.
-            var _ = Renderer.Dispatcher.InvokeAsync(() => Renderer.ProcessBufferedRenderBatches());
+            _ = Renderer.Dispatcher.InvokeAsync(() => Renderer.ProcessBufferedRenderBatches());
+        }
+
+        internal async Task EndInvokeJSFromDotNet(long asyncCall, bool succeded, string arguments)
+        {
+            try
+            {
+                AssertInitialized();
+
+                await Renderer.Dispatcher.InvokeAsync(() =>
+                {
+                    SetCurrentCircuitHost(this);
+                    if (!succeded)
+                    {
+                        // We can log the arguments here because it is simply the JS error with the call stack.
+                        Log.EndInvokeJSFailed(_logger, asyncCall, arguments);
+                    }
+                    else
+                    {
+                        Log.EndInvokeJSSucceeded(_logger, asyncCall);
+                    }
+
+                    DotNetDispatcher.EndInvoke(arguments);
+                });
+            }
+            catch (Exception ex)
+            {
+                Log.EndInvokeDispatchException(_logger, ex);
+            }
+        }
+
+        internal async Task DispatchEvent(string eventDescriptorJson, string eventArgs)
+        {
+            try
+            {
+                AssertInitialized();
+
+                await Renderer.Dispatcher.InvokeAsync(() =>
+                {
+                    SetCurrentCircuitHost(this);
+
+                    var eventDescriptor = ParseEventDescriptor(eventDescriptorJson);
+                    return RendererRegistryEventDispatcher.DispatchEvent(eventDescriptor, eventArgs);
+                });
+            }
+            catch (Exception ex)
+            {
+                Log.DispatchEventFailedToDispatchEvent(_logger, ex);
+                UnhandledException?.Invoke(this, new UnhandledExceptionEventArgs(ex, isTerminating: false));
+            }
+        }
+
+        private RendererRegistryEventDispatcher.BrowserEventDescriptor ParseEventDescriptor(string eventDescriptorJson)
+        {
+            try
+            {
+                return JsonSerializer.Deserialize<RendererRegistryEventDispatcher.BrowserEventDescriptor>(
+                    eventDescriptorJson,
+                    JsonSerializerOptionsProvider.Options);
+            }
+            catch (Exception ex)
+            {
+                Log.DispatchEventFailedToParseEventDescriptor(_logger, ex);
+                return null;
```

Comment:
Why return `null` here instead of throwing and letting the circuit be torn down? Won't this result in a confusing nullref exception from `RendererRegistryEventDispatcher.DispatchEvent`?

Reply:
We shouldn't throw simply because we got bad input as that would end up becoming a warning and we don't want warnings for user controlled inputs.

Reply:
But is returning `null` really correct? The consequence will be that we're passing a nonsense parameter to `RendererRegistryEventDispatcher.DispatchEvent`. Surely that's not what we want.

Reply:
I added a check for null and skipped invoking DispatchEvent on the calling method